ethereum/statetransition/src/main/java/tech/pegasys/artemis/statetransition/util/EpochProcessorUtil.java

```diff
@@ -135,6 +152,177 @@ public static void updateCrosslinks(BeaconState state) throws Exception {
     }
   }
 
+  /**
+   * Rewards and penalties applied with respect to justification and finalization. Spec:
+   * https://github.com/ethereum/eth2.0-specs/blob/v0.1/specs/core/0_beacon-chain.md#justification-and-finalization
+   *
+   * @param state
+   */
+  public static void justificationAndFinalization(BeaconState state) throws Exception {
+
+    final UnsignedLong FOUR = UnsignedLong.valueOf(4L);
+    UnsignedLong epochs_since_finality =
+        BeaconStateUtil.get_next_epoch(state).minus(state.getFinalized_epoch());
+    UnsignedLong previous_total_balance = previous_total_balance(state);
+    List<UnsignedLong> balances = state.getValidator_balances();
+
+    // Case 1: epochs_since_finality <= 4:
+    if (epochs_since_finality.compareTo(FOUR) <= 0) {
+      // Expected FFG source
+      UnsignedLong previous_balance =
+          AttestationUtil.get_previous_epoch_justified_attesting_balance(state);
+      List<Integer> previous_indices =
+          AttestationUtil.get_previous_epoch_justified_attester_indices(state);
+      case_one_penalties_and_rewards(
+          state, balances, previous_total_balance, previous_balance, previous_indices);
+
+      // Expected FFG target
+      previous_balance = AttestationUtil.get_previous_epoch_boundary_attesting_balance(state);
+      previous_indices = AttestationUtil.get_previous_epoch_boundary_attester_indices(state);
+      case_one_penalties_and_rewards(
+          state, balances, previous_total_balance, previous_balance, previous_indices);
+
+      // Expected beacon chain head
+      previous_balance = AttestationUtil.get_previous_epoch_head_attesting_balance(state);
+      previous_indices = AttestationUtil.get_previous_epoch_head_attester_indices(state);
+      case_one_penalties_and_rewards(
+          state, balances, previous_total_balance, previous_balance, previous_indices);
+
+      // Inclusion distance
+      UnsignedLong reward_delta = UnsignedLong.ZERO;
+      previous_indices = AttestationUtil.get_previous_epoch_attester_indices(state);
+      for (int index : previous_indices) {
+        UnsignedLong inclusion_distance = AttestationUtil.inclusion_distance(state, index);
+        reward_delta =
+            base_reward(state, index, previous_total_balance)
+                .times(UnsignedLong.valueOf(MIN_ATTESTATION_INCLUSION_DELAY))
+                .dividedBy(inclusion_distance);
+        apply_penalty_or_reward(balances, index, reward_delta, true);
+      }
+
+      // Case 2: epochs_since_finality > 4:
+    } else if (epochs_since_finality.compareTo(FOUR) > 0) {
+
+      Function<Integer, UnsignedLong> inactivity_penalty =
+          new Function<Integer, UnsignedLong>() {
```

Comment:
This might just be me being new to Java, but why define these functions here? seems a bit too complex and hard to read for me.

Reply:
Also, didn't know you could assign functions to variables and pass it around to other functions like in JS, super cool.

Reply:
I did this so that I could pass in a lambda function into `case_two_penalties_and_rewards` so that i didn't have to write the entire function 4x.  Not that different than python